### PR TITLE
fix: safeguard LastInterestTime against hard forks

### PR DESCRIPTION
**Pull request by toteki**
## Description

This will prevent the excessive (~10^20 factor) interest accrued during our hard fork of asherah testnet, where we modified LastInterestTime in genesis.

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] added appropriate labels to the PR
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [x] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -97,6 +97,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 - [987](https://github.com/umee-network/umee/pull/987) Streamline x/leverage CLI tests
 - [1012](https://github.com/umee-network/umee/pull/1012) Improve negative time elapsed error message
 - [1236](https://github.com/umee-network/umee/pull/1236) Improve leverage event fields.
+- [XXXX](https://github.com/umee-network/umee/pull/XXXX) Safeguards LastInterestTime against time reversals and unintended interest from hard forks.
```
Comment (toteki) [inline CHANGELOG.md:100]:
```suggestion
- [1288](https://github.com/umee-network/umee/pull/1288) Safeguards LastInterestTime against time reversals and unintended interest from hard forks.
```

File: x/leverage/keeper/interest.go [hunk 2]
```diff
@@ -80,9 +84,19 @@ func (k Keeper) AccrueAllInterest(ctx sdk.Context) error {
...
+		// if LastInterestTime appears to tbe in the future, do nothing (besides logging) and leave
+		// LastInterestTime at its stored value. This will repeat every block until BlockTime exceeds
+		// LastInterestTime.
 		return nil
 	}
+
 	yearsElapsed := sdk.NewDec(currentTime - prevInterestTime).QuoInt64(types.SecondsPerYear)
+	if yearsElapsed.GTE(sdk.OneDec()) {
+		// this safeguards primarily against misbehaving block time or incorrectly modified genesis states
+		// which would accrue significant interest on borrows instantly. Chain will halt.
```
Comment (RafilxTenfen) [inline x/leverage/keeper/interest.go:96]: Where the chain will halt?

Comment (toteki) [reply x/leverage/keeper/interest.go:96]: Error in `AccrueAllInterest`, which occurs during `EndBlock`, becomes a panic (similar to when we see negative time elapsed)

**Comment by RafilxTenfen** [comment]
can we backport this to 2.x.x ?

**Review by RafilxTenfen — APPROVED** [review]

File: x/leverage/keeper/interest.go (hunk 2 above)
Comment (robert-zaremba) [reply x/leverage/keeper/interest.go:96]: How it is possible that we can each such a long diff? The interest is accuring regularly.

File: x/leverage/keeper/interest.go [hunk 3]
```diff
@@ -178,6 +192,14 @@ func (k *Keeper) SetLastInterestTime(ctx sdk.Context, interestTime int64) error
 	store := ctx.KVStore(k.storeKey)
 	timeKey := types.CreateLastInterestTimeKey()
 
+	prevTime := k.GetLastInterestTime(ctx)
+
+	if interestTime < prevTime {
```
Comment (robert-zaremba) [inline x/leverage/keeper/interest.go:197]: when this is possible?

File: x/leverage/keeper/keeper_test.go [hunk 4]
```diff
@@ -661,6 +661,36 @@ func (s *IntegrationTestSuite) TestAccrueZeroInterest() {
...
+	// already has 1k u/umee for collateral.
+	addr, _ := s.initBorrowScenario()
+
+	// user borrows 40 umee
+	err := s.app.LeverageKeeper.Borrow(s.ctx, addr, sdk.NewInt64Coin(umeeapp.BondDenom, 40000000))
+	s.Require().NoError(err)
+
+	// Setting last interest time to a negative value is not allowed
+	err = s.tk.SetLastInterestTime(s.ctx, -1)
+	s.Require().Error(err)
```
Comment (robert-zaremba) [inline x/leverage/keeper/keeper_test.go:675]: let's check errors using `ErrorContains`

**Review by robert-zaremba — COMMENTED** [review]
Could you add a bit more comments? You are investigating cases which potentially are not possible. It's OK to make that check, but let's describe it for completness when these things can happen. Otherwise it's a bit confusing.... it's like checking `if 2+2 > 0`.

File: x/leverage/keeper/interest.go (hunk 3 above)
Comment (toteki) [reply x/leverage/keeper/interest.go:197]: It shouldn't be possible - but there's still that bug mentioned in the comments of AccrueAllInterest, where block time moves backwards. Protecting against it at the `Set` level seems to be a good redundant defense.

`// TODO fix this when tendermint solves https://github.com/tendermint/tendermint/issues/8773`

**Comment by toteki** [comment]
> How it is possible that we can each such a long diff? The interest is accruing regularly.

Scenarios where time difference of a year would be possible
- If a chain halt was so long, the next block would be a year later. (This is not what I expect)
- If `BlockTime` behaves incorrectly (Shouldn't happen, but bugs are always possible and this one would be _lethal_ to our users.)
- If `LastInterestTime` in genesis file was modified during a hard fork, and the next block would see massive interest accrue on all borrows to catch up (we suspect this just happened on Asherah-2).

Curiously, the `if lastInterestTime == 0` condition in `AccrueAllInterest` should have protected us, but it didn't.

**Comment by toteki** [comment]
> can we backport this to 2.x.x ?

Yeah (might take a few manual adjustments on the backport PR).

Also I need to add a gov proposal types to adjust interest scalar and set reserve amounts - those will be a separate PR but if we backport them rather than modifying genesis file, we could fix Asherah-2 with a binary swap + proposals instead of directly editing genesis. Might be faster to just edit the file though.

File: x/leverage/keeper/keeper_test.go (hunk 4 above)
Comment (toteki) [reply x/leverage/keeper/keeper_test.go]: Used `ErrIs(err,types.ErrNegativeTimeElapsed)` at `67fd679`

File: x/leverage/keeper/interest.go [hunk 5]
```diff
@@ -80,9 +84,19 @@ func (k Keeper) AccrueAllInterest(ctx sdk.Context) error {
 			"prev", prevInterestTime,
 		)
 
+		// if LastInterestTime appears to tbe in the future, do nothing (besides logging) and leave
```
Comment (toteki) [inline x/leverage/keeper/interest.go:87]:
```suggestion
		// if LastInterestTime appears to be in the future, do nothing (besides logging) and leave
```

File: x/leverage/keeper/interest.go (hunk 3 above)
Comment (robert-zaremba) [reply x/leverage/keeper/interest.go:197]: would be good to add a comment to clarify this here. Otherwise it looks strange.

**Pull request merged**